src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -924,177 +940,121 @@ private int RemoveStaleEntriesAndGetAvailableMetricPointRare(LookupData lookupDa
-            {
-                var shouldSample = this.exemplarFilter.ShouldSample(value, tags);
-                this.metricPoints[index].UpdateWithExemplar(value, tags: default, shouldSample);
-            }
-            else
-            {
-                this.metricPoints[index].Update(value);
-            }
-        }
-        catch (Exception)
-        {
-            Interlocked.Increment(ref this.DroppedMeasurements);
-            OpenTelemetrySdkEventSource.Log.MeasurementDropped(this.name, "SDK internal error occurred.", "Contact SDK owners.");
-        }
+        this.UpdateLongMetricPoint(index, value, tags: default);
     }
 
     private void UpdateLongCustomTags(long value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
     {
-        try
-        {
-            var index = this.FindMetricAggregatorsCustomTag(tags);
-            if (index < 0)
-            {
-                Interlocked.Increment(ref this.DroppedMeasurements);
+        var index = this.FindMetricAggregatorsCustomTag(tags);
 
-                if (this.EmitOverflowAttribute)
-                {
-                    this.InitializeOverflowTagPointIfNotInitialized();
-                    this.metricPoints[1].Update(value);
-                    return;
-                }
-                else
-                {
-                    if (Interlocked.CompareExchange(ref this.metricCapHitMessageLogged, 1, 0) == 0)
-                    {
-                        OpenTelemetrySdkEventSource.Log.MeasurementDropped(this.name, this.metricPointCapHitMessage, MetricPointCapHitFixMessage);
-                    }
+        this.UpdateLongMetricPoint(index, value, tags);
+    }
 
-                    return;
-                }
-            }
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void UpdateLongMetricPoint(int metricPointIndex, long value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
+    {
+        if (metricPointIndex < 0)
+        {
+            Interlocked.Increment(ref this.DroppedMeasurements);
 
-            // TODO: can special case built-in filters to be bit faster.
-            if (this.IsExemplarEnabled())
+            if (this.EmitOverflowAttribute)
             {
-                var shouldSample = this.exemplarFilter.ShouldSample(value, tags);
-                this.metricPoints[index].UpdateWithExemplar(value, tags, shouldSample);
+                this.InitializeOverflowTagPointIfNotInitialized();
+                this.metricPoints[1].Update(value);
             }
-            else
+            else if (Interlocked.CompareExchange(ref this.metricCapHitMessageLogged, 1, 0) == 0)
             {
-                this.metricPoints[index].Update(value);
+                OpenTelemetrySdkEventSource.Log.MeasurementDropped(this.name, this.metricPointCapHitMessage, MetricPointCapHitFixMessage);
             }
+
+            return;
         }
-        catch (Exception)
+
+        switch (this.exemplarFilter)
```

Comment:
It would probably be faster to use `if` than `switch`? If yes, we should optimize the non-exemplar updates.

Reply:
Found an old benchmark saying that switch usually will be faster cause compiler will create jump table for switch statement: https://www.blackwasp.co.uk/SpeedTestIfElseSwitch.aspx.
I _guess_ the assembly generated for if/switch here will be identical given there are only 3 cases.

Reply:
If you know your most common condition, you can make that the very first check. That way you only have to check `if(true)` to execute the most common path. That might be faster than the jump table lookup. In fact, you can use a mixed approach as well. Something like this:

```csharp
if (mostFrequentlyHitCondition)
{
   RunSomething();
}

switch (lessFrequentlyHitConditions)
{
case Condition1:

case Condition2:

...

default:

}
```

Reply:
I updated the code to use an `if` block. But it has been interesting to try and prove what is best.

Here are some benchmarks run without PGO enabled:

| Method                  | ExemplarFilterType | Mean     | Error     | StdDev    | Allocated |
|------------------------ |------------------- |---------:|----------:|----------:|----------:|
| UpdateOldStyle          | AlwaysOff          | 2.458 ns | 0.0722 ns | 0.0676 ns |         - |
| UpdateInlineSwitchStyle | AlwaysOff          | 1.572 ns | 0.0366 ns | 0.0342 ns |         - |
| UpdateInlineIfStyle     | AlwaysOff          | 1.353 ns | 0.0457 ns | 0.0626 ns |         - |
| UpdateOldStyle          | AlwaysOn           | 2.938 ns | 0.0206 ns | 0.0172 ns |         - |
| UpdateInlineSwitchStyle | AlwaysOn           | 1.800 ns | 0.0343 ns | 0.0304 ns |         - |
| UpdateInlineIfStyle     | AlwaysOn           | 1.800 ns | 0.0366 ns | 0.0342 ns |         - |
| UpdateOldStyle          | TraceBased         | 4.666 ns | 0.0629 ns | 0.0588 ns |         - |
| UpdateInlineSwitchStyle | TraceBased         | 3.297 ns | 0.0470 ns | 0.0417 ns |         - |
| UpdateInlineIfStyle     | TraceBased         | 3.198 ns | 0.0382 ns | 0.0319 ns |         - |

Here are some benchmarks run with PGO enabled:

| Method                  | ExemplarFilterType | Mean     | Error     | StdDev    | Allocated |
|------------------------ |------------------- |---------:|----------:|----------:|----------:|
| UpdateOldStyle          | AlwaysOff          | 1.299 ns | 0.0410 ns | 0.0383 ns |         - |
| UpdateInlineSwitchStyle | AlwaysOff          | 1.345 ns | 0.0374 ns | 0.0313 ns |         - |
| UpdateInlineIfStyle     | AlwaysOff          | 1.106 ns | 0.0316 ns | 0.0296 ns |         - |
| UpdateOldStyle          | AlwaysOn           | 1.359 ns | 0.0141 ns | 0.0132 ns |         - |
| UpdateInlineSwitchStyle | AlwaysOn           | 1.373 ns | 0.0514 ns | 0.0481 ns |         - |
| UpdateInlineIfStyle     | AlwaysOn           | 1.325 ns | 0.0310 ns | 0.0275 ns |         - |
| UpdateOldStyle          | TraceBased         | 2.756 ns | 0.0709 ns | 0.0663 ns |         - |
| UpdateInlineSwitchStyle | TraceBased         | 3.392 ns | 0.0452 ns | 0.0422 ns |         - |
| UpdateInlineIfStyle     | TraceBased         | 2.234 ns | 0.0231 ns | 0.0204 ns |         - |

PGO is smart enough to see that one value is used repeatedly and optimize the code on its own for that value. It is great for what we're doing here.

I ended up going with the `if` because PGO may not be there in AOT deployments/older runtimes so it seems best to optimize for it not being there.